src/hip_stream.cpp

```diff
@@ -254,14 +254,131 @@ hipError_t hipStreamGetPriority(hipStream_t stream, int* priority) {
+
+    tempBarrier = new hsa_barrier_and_packet_t;
+    memset(tempBarrier, 0, sizeof(hsa_barrier_and_packet_t));
+    tempBarrier->header = HSA_PACKET_TYPE_INVALID;
+
+    // Barrier 1
+    *barrier1 = &(((hsa_barrier_and_packet_t*)(queue->base_address))[index & mask]);
+    memcpy(*barrier1,tempBarrier,sizeof(hsa_barrier_and_packet_t));
+
+    // Barrier 2
+    *barrier2 = &(((hsa_barrier_and_packet_t*)(queue->base_address))[nextIndex & mask]);
+    memcpy(*barrier2,tempBarrier,sizeof(hsa_barrier_and_packet_t));
+
+    delete tempBarrier;
+
+    return hipSuccess;
+}
+
 hipError_t hipStreamAddCallback(hipStream_t stream, hipStreamCallback_t callback, void* userData,
                                 unsigned int flags) {
+
     HIP_INIT_API(hipStreamAddCallback, stream, callback, userData, flags);
     hipError_t e = hipSuccess;
 
-    // Create a thread in detached mode to handle callback
+    if(stream == hipStreamNull)
+    {
+        ihipCtx_t* device = ihipGetTlsDefaultCtx();
+        stream = device->_defaultStream;
+    }
+
+    // 1. Lock the queue
+    hsa_queue_t* lockedQ = static_cast<hsa_queue_t*> (stream->criticalData()._av.acquire_locked_hsa_queue());
+
+    if(lockedQ == nullptr)
+    {
+        // No queue attached to stream hence exiting early
+        return ihipLogStatus(hipErrorMissingConfiguration);
+    }
+
+    // 2. Allocate a singals
+    hsa_signal_t signal;
+    hsa_status_t status = hsa_signal_create(1, 0, NULL, &signal);
+
+    if(status != HSA_STATUS_SUCCESS)
+    {
+        return ihipLogStatus(hipErrorInvalidValue);
+    }
+
+    hsa_signal_t depSignal;
+    status = hsa_signal_create(1, 0, NULL, &depSignal);
+
+    if(status != HSA_STATUS_SUCCESS)
+    {
+        return ihipLogStatus(hipErrorInvalidValue);
+    }
+
+    // 3. Store callback details
     ihipStreamCallback_t* cb = new ihipStreamCallback_t(stream, callback, userData);
-    std::thread(ihipStreamCallbackHandler, cb).detach();
+    if(cb == nullptr)
+    {
+        return ihipLogStatus(hipErrorMemoryAllocation);
+    }
+    cb->_signal = depSignal;
+
+    // 4. Create barrier packets
+    uint64_t index ;
+    uint64_t nextIndex;
+
+    hsa_barrier_and_packet_t* barrier;
+    hsa_barrier_and_packet_t* depBarrier;
+
+    setCallbackPacket(lockedQ, index, nextIndex, &barrier, &depBarrier);
+
+    barrier->completion_signal = signal;
+
+    depBarrier->dep_signal[0] = depSignal;
+
+    uint16_t header = (HSA_PACKET_TYPE_BARRIER_AND << HSA_PACKET_HEADER_TYPE)| 1 << HSA_PACKET_HEADER_BARRIER;
+
+    // 5. Update packet header,
+    // Intentionally updated second barrier header before first in order to avoid race
+    depBarrier->header = header;
+    barrier->header = header;
+
+    // 6. Trigger the doorbell
+    nextIndex = nextIndex + 1;
+    hsa_queue_store_write_index_screlease(lockedQ, nextIndex);
+    hsa_signal_store_relaxed(lockedQ->doorbell_signal, index+1);
+
+    // 7. Release queue
+    stream->criticalData()._av.release_locked_hsa_queue();
```

Comment:
Based on my previous request to lock the stream first in the above code then then the stream needs to be unlocked here as well.

Reply:
Thanks for pointing out added the change